### ISSUE TITLE
Some weirdness on Mac OS Catalina **[BUG]**

### ISSUE THREAD

**Issue by linosia97** [CGCookie/retopoflow#753]
**My familiarity with the project is as follows (check one):**
- [ ] I just started using this project.

**Retopoflow Version:** 3.0.0α

**Blender Version (Hash):** 2.81.16 master 2019-11-20

**Platform:** Mac OS Catalina 10.15.1

**Hardvare** MacBook Air 2017 I5 5350U, Intel HD 6000

**Issue:**
Weird graphical glitches, crashes when opened in full-app window (like a separate desktop).

{screenshot of issue
![Screenshot 2019-11-26 at 19 02 02](https://user-images.githubusercontent.com/49729502/69655892-30466700-1080-11ea-8327-7f76e8820618.png)
![Screenshot 2019-11-26 at 18 58 38](https://user-images.githubusercontent.com/49729502/69655901-33415780-1080-11ea-9fc9-563124185f67.png)

}

**Replication Resources:**

**Replication Steps:**

1. Just install plugin and go to retopoflow menu.

**Comment by linosia97** [comment]
Environment:

- RetopoFlow: 3.0.0α
- Blender: 2.81.16 master 2019-11-20
- Platform: Darwin, 19.0.0, Darwin Kernel Version 19.0.0: Thu Oct 17 16:17:15 PDT 2019; root:xnu-6153.41.3~29/RELEASE_X86_64, x86_64, i386
- GPU: Intel Inc., Intel(R) HD Graphics 6000, 4.1 INTEL-14.2.16, 4.10
- Timestamp: 2019-11-26 19:01:41.274281
- Undo: 

Error Hash: 6dda22a701aa352d035a9ca96c179158

Trace:

- EXCEPTION (<class 'RuntimeError'>): Error: Cannot open file RetopoFlow_backup.blend@ for writing: Read-only file system
- 
-     /Applications/Blender.app/Contents/Resources/2.81/scripts/modules/bpy/ops.py
- 000 0201:__call__() ret = op_call(self.idname_py(), None, kw)
-     /Users/linosia/Library/Application Support/Blender/2.81/scripts/addons/retopoflow/retopoflow/rf/rf_blender.py
- 001 0391:save_backup() bpy.ops.wm.save_as_mainfile(filepath=filepath, check_existing=False, copy=True)
- 002 0368:check_auto_save() self.save_backup()
-     /Users/linosia/Library/Application Support/Blender/2.81/scripts/addons/retopoflow/retopoflow/rf/rf_states.py
- 003 0124:modal_main() self.check_auto_save()
-     .../common/fsm.py
- 004 0063:run() return fn(*args, **kwargs)

**Comment by linosia97** [comment]
Also, it's quite slow to use it...

**Comment by vxlcoder** [comment]
Thanks for the report, @linosia97.

The actual exception should be an easy fix... will work on that ASAP.

As for the visual artifacts... I have never seen this before, and I have no idea what could be causing it.  Very strange!

Does it always look like that, or does it start looking like that after doing something?

Does it flicker, especially when you rotate the view around (MMB drag)?  Any chance that you can do a video capture?

Have you had these effects show up in any other application?

**Comment by linosia97** [comment]
Sorry for late reply! Will upload video in 10-15min

**Comment by vxlcoder** [comment]
That video will be super helpful, thanks! I might also have you try out a
few other things, too.

The original issue (unable to open the backup file) should now be fixed.
Please let me know if you continue to have issues after you update.

**Comment by linosia97** [comment]
https://www.youtube.com/watch?v=p_UbqHmU2A8&feature=youtu.be
Here is the video

**Comment by linosia97** [comment]
I should comment that I DID NOT cloned it through GitHub like I should

**Comment by vxlcoder** [comment]
Oh! I think I might know the problem... one sec

**Comment by vxlcoder** [comment]
what are your `clip start` and `clip end` set to?

![image](https://user-images.githubusercontent.com/3790699/69814778-f2059080-11c2-11ea-8e18-3dcdcafc3d10.png)

**Comment by linosia97** [comment]
Yes)

**Comment by vxlcoder** [comment]
sorry. what number is in for the clip start and for clip end?  is it same as mine (0.5m and 5000m)?

**Comment by linosia97** [comment]
Yes

**Comment by linosia97** [comment]
![Screenshot 2019-11-28 at 16 43 15](https://user-images.githubusercontent.com/49729502/69815061-319d9e00-11fe-11ea-96ae-fe25ee2db052.png)

**Comment by vxlcoder** [comment]
hmmm. ok.  new idea... i'm wondering if it's a problem with the position/resolution in the shader.  will take a look.

thank you for helping me debug this!

**Comment by linosia97** [comment]
Also, for some reason Blender with retopoflow in background consumes A LOT of CPU power.
Also, after entering and exiting retopoflow scale of default cube becomes 5 instead of 1 as it should remain...

**Comment by linosia97** [comment]
![Screenshot 2019-11-28 at 16 46 45](https://user-images.githubusercontent.com/49729502/69815404-ea63dd00-11fe-11ea-9b78-5ea8dc6058af.png)
![Screenshot 2019-11-28 at 16 46 18](https://user-images.githubusercontent.com/49729502/69815405-ea63dd00-11fe-11ea-9938-9f5ad79ab32b.png)

**Comment by linosia97** [comment]
Well, with CPU power seems just some glitch...

But scaling is changing from 1 to 5, which isn't normal I think...

**Comment by linosia97** [comment]
Also, does tools should work now or they will be working later? (I am a newbie and don't know how to use it at all)
And when to expect .zip-ready download from GitHub?

**Comment by vxlcoder** [comment]
all the source and target meshes are scaled to certain size so that all of the rendering settings work properly.  i'm still working on this, but you shouldn't even notice once it is done (RF will put everything back to how it was).

CPU usage may jump a bit when you start or end RF or when you move the mouse around.  But this shouldn't be significantly more than normal use.  if you can reproduce it, let me know (start a new issue), and I'll work on that.

a nicely packaged zip is available on the Blender Market, or you can try downloading it directly from GH (require getting it in parts: RF and CookieCutter).  not sure when we'll have a packaged zip on GH, yet.  will post a link to a zip if I have a test version for you, though.

**Comment by linosia97** [comment]
Ok, thank you for your hard work ;)

**Comment by vxlcoder** [comment]
I'm going to have you try something... I don't think this will fix it, but let's try anyways.

Can you open a file in a text editor.  The file is located in `your_blender_path/2.81/scripts/addons/RetopoFlow/addon_common/common/shaders/ui_element.glsl`.  Open the file, and then remove the second line, which should be `uniform vec2 screen_size;`, save, then run Blender.

OSX has always been _very_ particular about their shaders.  the `screen_size` variable isn't used, but maybe it's causing some weird problem... not sure.

**Comment by linosia97** [comment]
Tried, but nope -- same issue

**Comment by vxlcoder** [comment]
ok. i have one other test.  again, i don't think this will work, but just a test

remove current RF version, download and install this one https://www.dropbox.com/s/sjklq70oyvf5h25/RetopoFlow_v3.00.0.zip?dl=0

**Comment by linosia97** [comment]
Again, same issue...

**Comment by vxlcoder** [comment]
ok, let's try one more thing... (again, thank you for helping debug this!)

open `RetopoFlow/addon_common/common/drawing.py`

starting line 401, add the `print(last_m)` line as in the following:

```
    def get_pixel_matrix(self):
        '''
        returns MVP for pixel view
        TODO: compute separate M,V,P matrices
        '''
        last_w, last_h, last_m = -1, -1, None
        def fn():
            nonlocal last_w, last_h, last_m
            if not self.r3d: return None
            w,h = self.rgn.width,self.rgn.height
            if last_w != w or last_h != h:
                last_w,last_h = w,h
                last_m = Matrix([[2/w,0,0,-1], [0,2/h,0,-1], [0,0,1,0], [0,0,0,1]])
                print(last_m)  # <------- add this line here
            return last_m
        self.get_pixel_matrix = fn
        return fn()
```

note: there are 16 spaces on the left side of the `print`

**Comment by vxlcoder** [comment]
also, in the zip I gave you, does the original uncaught exception issue show back up?

**Comment by vxlcoder** [comment]
oh, forgot to mention this... can you run Blender from the terminal, because this will print details to your terminal that you'll need to copy here.  if you don't know how, let me know, and we'll try something different

**Comment by linosia97** [comment]
>also, in the zip I gave you, does the original uncaught exception issue show back up?
It's gone.

**Comment by linosia97** [comment]
![Screenshot 2019-11-28 at 17 31 59](https://user-images.githubusercontent.com/49729502/69818366-4598ce00-1205-11ea-8091-35392e4fba5c.png)
Why this arrows flickers when I do something? (like rotating object in viewport...)

**Comment by linosia97** [comment]
>ok, let's try one more thing

Nope, same issue

**Comment by linosia97** [comment]
>oh, forgot to mention this... can you run Blender from the terminal, because this will print details to your terminal that you'll need to copy here. if you don't know how, let me know, and we'll try something different

Can you give me a command that I should write in terminal? (I know just one for opening new window: open -na Blender)

**Comment by vxlcoder** [comment]
"it's gone" great!

"flicker" unfortunately, that's a blender issue.  i'll need to talk to the devs about it.

"same issue" do you know how to run Blender from the terminal?  (this is for older blender, but should still work: https://blenderartists.org/t/start-blender-2-57-from-mac-os-x-terminal-with-python-output/508418)  you'll need to run Blender from terminal, because RF will print useful debug info to terminal window.

**Comment by linosia97** [comment]
[retopoflowBlenderFromTerminal.docx](https://github.com/CGCookie/retopoflow/files/3902418/retopoflowBlenderFromTerminal.docx)

**Comment by linosia97** [comment]
This is all I got from terminal...

**Comment by linosia97** [comment]
the beginning of it:

Last login: Thu Nov 28 17:29:03 on ttys000

The default interactive shell is now zsh.
To update your account to use zsh, please run `chsh -s /bin/zsh`.
For more details, please visit https://support.apple.com/kb/HT208050.
MacBook-Air-Linosia:~ linosia$ /Applications/Blender.app/Contents/MacOS/Blender 
Read prefs: /Users/linosia/Library/Application Support/Blender/2.81/config/userpref.blend
found bundled python: /Applications/Blender.app/Contents/Resources/2.81/python
RetopoFlow git: None
RetopoFlow options path: /Users/linosia/Library/Application Support/Blender/2.81/scripts/addons/RetopoFlow/RetopoFlow_options.json
RetopoFlow: self.update_displace_option()
Relax reset <RetopoFlow.retopoflow.rfwidgets.rfwidget_brushfalloff.RFWidget_BrushFalloff_Relax object at 0x119228f10>
Loaded font "/Users/linosia/Library/Application Support/Blender/2.81/scripts/addons/RetopoFlow/addon_common/common/fonts/DroidSans-Blender.ttf" as id 3
Loading image "contours_32.png" (/Users/linosia/Library/Application Support/Blender/2.81/scripts/addons/RetopoFlow/addon_common/common/../../icons/contours_32.png)
Loading image "polystrips_32.png" (/Users/linosia/Library/Application Support/Blender/2.81/scripts/addons/RetopoFlow/addon_common/common/../../icons/polystrips_32.png)
Loading image "polypen_32.png" (/Users/linosia/Library/Application Support/Blender/2.81/scripts/addons/RetopoFlow/addon_common/common/../../icons/polypen_32.png)
Loading image "relax_32.png" (/Users/linosia/Library/Application Support/Blender/2.81/scripts/addons/RetopoFlow/addon_common/common/../../icons/relax_32.png)
Loading image "tweak_32.png" (/Users/linosia/Library/Application Support/Blender/2.81/scripts/addons/RetopoFlow/addon_common/common/../../icons/tweak_32.png)
Loading image "loops_32.png" (/Users/linosia/Library/Application Support/Blender/2.81/scripts/addons/RetopoFlow/addon_common/common/../../icons/loops_32.png)
Loading image "patches_32.png" (/Users/linosia/Library/Application Support/Blender/2.81/scripts/addons/RetopoFlow/addon_common/common/../../icons/patches_32.png)
Loading image "strokes_32.png" (/Users/linosia/Library/Application Support/Blender/2.81/scripts/addons/RetopoFlow/addon_common/common/../../icons/strokes_32.png)
Loading image "collapse_close.png" (/Users/linosia/Library/Application Support/Blender/2.81/scripts/addons/RetopoFlow/addon_common/common/images/collapse_close.png)

**Comment by vxlcoder** [comment]
thanks.  everything looks fine (although, there is another, unrelated bug...).

i might have to think a bit more on this.

**Comment by linosia97** [comment]
Glad to help ;)

**Comment by linosia97** [comment]
Any progress with Mac OS?
Because it's 2-nd beta now, and the interface still buggy :)
(also notice some unwanted checker-like texture on greyed out area, outside retopoflow)
Macbook air 2017, Mac OS Catalina 10.15.4 (latest, latest drivers).
Blender 2.83 beta
![Screenshot 2020-05-10 at 01 09 54](https://user-images.githubusercontent.com/49729502/81486115-3f4e5980-925b-11ea-973c-ab8032add83d.png)

**Comment by linosia97** [comment]
Also, big thanks to including "addon_common" into b280 tree -- now it's muuuuch easier to install new version of Retopoflow!

**Comment by vxlcoder** [comment]
hmmm, that's still very strange @linosia97.  I do have a macbook pro running 10.15.4 with Nvidia GeForce GT 750M 2GB and Blender 2.83beta (just downloaded it now) and I'm not seeing anything close to this.  pretty sure it's a driver issue, especially since the issue persists across blender versions and major changes to RF rendering.

just checking, is this the very latest of the `b280` branch?  (i've made some significant changes recently)

does this only appear around the UI?  does it show around the target rendering?  (try using the tools to see what it looks like)

Thanks for helping us debug this!

**Comment by vxlcoder** [comment]
example of possibly being driver / card related: https://developer.blender.org/T73783

not exactly the same, but an indication

**Comment by linosia97** [comment]
>just checking, is this the very latest of the b280 branch? (i've made some significant changes recently)

Yes. Downloaded this 2 days ago :)

**Comment by linosia97** [comment]
>hmmm, that's still very strange @linosia97. I do have a macbook pro running 10.15.4 with Nvidia GeForce GT 750M 2GB and Blender 2.83beta (just downloaded it now) and I'm not seeing anything close to this. pretty sure it's a driver issue, especially since the issue persists across blender versions and major changes to RF rendering.

Remember -- I use **_integrated_** iGPU Intel HD 6000.
And yeah -- this causes a lot of issues you cannot see with discrete GPU

**Comment by linosia97** [comment]
>does this only appear around the UI? does it show around the target rendering? (try using the tools to see what it looks like)
Tools looks fine (I don't know how to actually use instruments yet 😅)
But yes -- only UI and helping window

**Comment by linosia97** [comment]
> example of possibly being driver / card related: https://developer.blender.org/T73783
> 
> not exactly the same, but an indication

And yes -- this is MY bug report)

**Comment by vxlcoder** [comment]
ha!  sorry, i've been distracted and running triage for the past few months... missed that :)

ok.  i think the workaround will have to be to turn off the UI (press F9 to toggle) and use shortcuts as much as possible.

recent change: switch tools by using the numbers across top of keyboard (1-contours, 2-polystrips, etc.).

presently: CTRL+LMB will perform an insert with tool.

hmmm... i wonder if this is an issue with alpha channel.  let me make work on a alpha-less version (workaround checkbox) to see if this helps.  by the way, do you use Blender's region overlap?

**Comment by linosia97** [comment]
>do you use Blender's region overlap?

I don't even know what it is)

>let me make work on a alpha-less version (workaround checkbox) to see if this helps

Ok, will test it)

**Comment by vxlcoder** [comment]
how comfortable are you in trying to edit a file?

**Comment by vxlcoder** [comment]
try this: download the latest version of RF and addon_common and install as usual.  then, before you start Blender, open the file `addon_common/common/shaders/ui_element.glsl` and modify line 39 to be `const bool DEBUG_IGNORE_ALPHA = true;`.  now, start Blender and RF.  if you can, post a screenshot.

**Comment by vxlcoder** [comment]
(you can also modify the file before zipping and installing in Blender)

**Comment by linosia97** [comment]
Downloaded 10 min ago
I did everything you wrote.
Nope -- same issue
![Screenshot 2020-05-19 at 20 43 49](https://user-images.githubusercontent.com/49729502/82360192-afa26b00-9a11-11ea-9c8d-345bf187883b.png)

**Comment by linosia97** [comment]
Also you placed tag "can't replicate".
Did you tested it with Mac OS (Mojave-Catalina) on Intel HD ?
Even if it's Mac OS with discrete (AMD?) card -- the result may be different...

**Comment by vxlcoder** [comment]
hmm.  does it do the same in b282a (or any other b28x?) or b290?

i just updated the .zip file I linked you in the email with the latest set of commits.

The "can't replicate" tag is a note for me, that I haven't been able to replicate this same error, not that it's not a non-reproducible bug.  i do have a mac, but it has a different card and therefore doesn't produce these visual errors.

**Comment by linosia97** [comment]
>i just updated the .zip file I linked you in the email with the latest set of commits.
Where exactly I can see this file?
I downloaded latest from GitHub from b280 branch

**Comment by vxlcoder** [comment]
> I downloaded latest from GitHub

excellent!  never mind about the .zip file comment.

does this visual bug show up in other versions of blender?

**Comment by linosia97** [comment]
ok, will download 2.82a and 2.90 and will test it

**Comment by vxlcoder** [comment]
thanks!

**Comment by vxlcoder** [comment]
i have a feeling that it will behave the same, as this is likely a HW-related issue, but I'd like to test.  thank you for helping!

**Comment by linosia97** [comment]
you are welcome :)

>HW-related issue
I don't know what it is :)
Like, rendering or viewport?

**Comment by vxlcoder** [comment]
sorry, HW=hardware.  could be driver or physical card.  very difficult to understand the problem without having first-hand experience.  but honestly, even if i were able to replicate here, a solution might still be difficult.

**Comment by linosia97** [comment]
tested with 2.82a and 2.90 with clean (not modified) latest retopoflow.
Same effect, and also slow UI (NOT tools, but UI around them).
![2 82a](https://user-images.githubusercontent.com/49729502/82368417-1463c280-9a1e-11ea-9a0c-c3f663b92d60.png)
![2 82a_2](https://user-images.githubusercontent.com/49729502/82368428-175eb300-9a1e-11ea-9bff-bd49cc5e151d.png)
![2 90](https://user-images.githubusercontent.com/49729502/82368435-19287680-9a1e-11ea-98dc-f477b230beda.png)

**Comment by linosia97** [comment]
will also test with 2.83 on windows 1909 (bootcamp, same hardware) to determine if it's really driver-only issue (previous time on windows it worked fine - no artifacts)

**Comment by linosia97** [comment]
Tested on windows 10 1909 (bootcamp) with blender 2.83 beta -- surprisingly same issue!
![Annotation 2020-05-19 233334](https://user-images.githubusercontent.com/49729502/82376948-90b0d280-9a2b-11ea-988d-326521863aaa.png)

**Comment by vxlcoder** [comment]
very strange.  i might try simplifying the shader used for all UI stuff to see if that clears up the issue...

**Comment by vxlcoder** [comment]
ok, @linosia97.  I have a possible workaround for this bug worked into commit a4a7657.  If you get a chance, can you try using the very latest version of RF, and let me know whether it worked or did not?  Thank you for your patience on this perplexing bug!

**Comment by linosia97** [comment]
hi, vxlcoder!
nope, same issue...
![Screenshot 2020-11-24 at 23 00 52](https://user-images.githubusercontent.com/49729502/100150966-26309180-2ea9-11eb-82a4-70df74508238.png)
to be honest, as of now I don't use blender as often, but still tested with fresh blender 2.91 and with commit you send here.

**Comment by vxlcoder** [comment]
hmmm... that's just too bad that it didn't work.  Thank you for trying it again!  will continue to work on this

**Comment by jlampel** [comment]
Closing only because this is a duplicate. Investigation will continue in https://github.com/CGCookie/retopoflow/issues/879 !

**Comment by vxlcoder** [comment]
for those with this issue, can you please try the `v3.2.1` branch to see whether the committed changes improve this visual bug?  see [link](https://docs.retopoflow.com/addon_updater.html) for details.

also, if possible, can someone post a video capture of using RetopoFlow with this visual issue?  I have a few other hypotheses that I think a video might provide the right evidence.

in particular, can you capture the following in the video:

- start with default cube
- start RF
- move the main/poly count/options windows around the screen slowly
- try minimizing / restoring each of the windows
- create new geometry using one of the tools
- rotate the view around just a bit
- open the help menu
- slowly scroll through the help
- try resizing the entire blender window very slowly

i think doing the items above might expose what the issue _might_ be.  thank you!!

**Comment by vxlcoder** [comment]
for those with this issue, can you please try the `v3.2.1` branch to see whether the committed changes improve this visual bug?  see [link](https://docs.retopoflow.com/addon_updater.html) for details.

also, if possible, can someone post a video capture of using RetopoFlow with this visual issue?  I have a few other hypotheses that I think a video might provide the right evidence.

in particular, can you capture the following in the video:

- start with default cube
- start RF
- move the main/poly count/options windows around the screen slowly
- try minimizing / restoring each of the windows
- create new geometry using one of the tools
- rotate the view around just a bit
- open the help menu
- slowly scroll through the help
- try resizing the entire blender window very slowly

i think doing the items above might expose what the issue _might_ be.  thank you!!

**Comment by vxlcoder** [comment]
Hi all!

I think I might have found the cause of this issue (and other similar issues), and now we have a solution!

Please, if you're still having this issue with RetopoFlow v3.2.2, follow the steps on using the [RetopoFlow Updater System](https://docs.retopoflow.com/addon_updater.html) to try out commit c1d4f0f.

IMPORTANT: be sure to start with v3.2.2 before updating to the commit!  the previous updater system had a bug that was fixed in v3.2.2.

Please let me know here if you

- continue having this issue even after checking out commit c1d4f0f, or
- no longer experience this issue after checkout out commit c1d4f0f, or
- were experiencing this issue before but it stopped being an issue even before checking out commit c1d4f0f.

Thanks!

**Comment by vxlcoder** [comment]
Hi all!  I have made progress on another (related?) issue #915 that might affect this issue.

Can you try using the updater system ([doc](https://docs.retopoflow.com/addon_updater.html)) to checkout branch `v3.2.4` to see if this change fixes this bug on your machine, and then report what happened / didn't happen here?

Thanks!
